python/cudaq/kernel/kernel_decorator.py

```diff
@@ -460,6 +472,12 @@ def __call__(self, *args):
                 mlirType = self.argTypes[i]
                 continue
 
+            if cc.StructType.isinstance(mlirType):
+                if cc.StructType.isinstance(self.argTypes[i]):
+                    memberTys = cc.StructType.getTypes(mlirType)
+                    if len(memberTys) == 0:
+                        mlirType = self.argTypes[i]
```

Comment:
I'm kind of confused by this loop. There are cases above here that set `mlirType` and then do a continue. But the loop will just advance to the next argument and its type. So what is going on there? Why set `mlirType` when it will be overwritten with the next argument type?

In this case, we reset the `mlirType` to whatever it was before we converted it to an MLIR type if and only if it has no members. And then we fall through to the rest of this case logic. The obvious question is why are we doing that? Is there some sort of product type that gets converted incorrectly? What does undoing the conversion to an MLIR type actually intending to do?

Reply:
The function is trying to cast actuals to formal types, so it sets the type of actual to the type of formal when the cast succeeds. The confusing part in struct type case is the check for the zero length; I will add a comment to clarify. This is a workaround for an issue I don't know how to solve properly yet - in a case of a direct call, actual tuple argument type is just "tuple", with no other details available. I converted it to a tuple of 0 length in `mlirTypeFromPyType` and allow it to be casted here to any tuple. I need to do some investigation to see how it is handed in `sample` and `run` cases which seem to show types with more information.